File: self-evolution/runs/SE-012/ui_probe.py
"""SE-012 UI probe: drive the built console against a fully stubbed /api so no AWS
is touched. Asserts: EDIT opens the inline form pre-filled from GET detail; SAVE is
disabled with a reason until a field changes and is valid; the confirm dialog names
the referencing agents + the expiry impact; PUT carries only the changed field; the
row refreshes from the response without a list reload."""
import json, sys
from playwright.sync_api import sync_playwright

BASE = "http://127.0.0.1:4199"
DEFAULT = "launchpad_memory-ABC123"
TEAM = "team_notes-XYZ789"
puts = []
list_calls = []
unhandled = []

def row(id_, name, default, agents, status="ACTIVE"):
    return {"id": id_, "arn": f"arn:mem:{id_}", "name": name, "status": status,
            "created_at": "2026-08-30T12:00:00+00:00", "updated_at": None,
            "is_default": default, "agents": agents}

def detail(id_, name, desc, days, status="ACTIVE", updated=None):
    return {"id": id_, "arn": f"arn:mem:{id_}", "name": name, "description": desc,
            "status": status, "failure_reason": None, "event_expiry_days": days,
            "execution_role_arn": None, "created_at": "2026-08-30T12:00:00+00:00",
            "updated_at": updated, "is_default": id_ == DEFAULT, "strategies": [],
            "namespace_keys": []}

def handle(route, request):
    url = request.url.split("?")[0]
    path = url[url.index("/api/"):]
    m = request.method
    body = None
    if path == "/api/auth/status":
        body = {"auth_required": False, "authenticated": True, "registration_enabled": False,
                "registration_requires_approval": False, "username": "admin", "role": "admin",
                "email": None, "account_expires_at": None, "permissions": []}
    elif path == "/api/workspaces" and m == "GET":
        body = {"workspaces": [{"id": "default", "name": "default", "account_id": "123456789012",
                "region": "us-west-2", "cross_account": False, "bootstrap_status": "ready",
                "is_default": True, "created_at": None, "updated_at": None}], "all_workspaces": True}
    elif path == "/api/memory/overview":
        body = {"configured": True, "memory": detail(DEFAULT, "launchpad_memory", None, 30),
                "strategies": [], "actor_count": 0, "actor_count_truncated": False,
                "other_memories": []}
    elif path == "/api/memory/resources" and m == "GET":
        list_calls.append(1)
        body = {"items": [row(DEFAULT, "launchpad_memory", True, []),
                          row(TEAM, "team_notes", False,
                              [{"id": "a1", "name": "hr-assistant"}, {"id": "a2", "name": "faq-bot"}])],
                "default_id": DEFAULT}
    elif path == f"/api/memory/resources/{TEAM}" and m == "GET":
        body = detail(TEAM, "team_notes", "per-team memory", 30)
    elif path == f"/api/memory/resources/{TEAM}" and m == "PUT":
        puts.append(json.loads(request.post_data))
        body = detail(TEAM, "team_notes", "per-team memory", 14, status="UPDATING",
                      updated="2026-09-05T09:30:00+00:00")
    else:
        unhandled.append((m, path))
        body = {}
    route.fulfill(status=200, content_type="application/json", body=json.dumps(body))

def check(cond, msg):
    print(("PASS " if cond else "FAIL ") + msg)
    if not cond:
        raise SystemExit(1)

with sync_playwright() as p:
    b = p.chromium.launch(args=["--no-sandbox"])
    page = b.new_page(viewport={"width": 1280, "height": 900})
    page.route("**/api/**", handle)
    page.goto(f"{BASE}/memory?view=resources")
    edits = page.get_by_test_id("memory-resource-edit")
    edits.nth(1).wait_for()
    check(edits.count() == 2, "EDIT action on every row (incl. the platform default)")
    edits.nth(1).click()
    er = page.get_by_test_id("memory-resource-edit-row")
    er.wait_for()
    desc = page.locator("#mem-res-edit-desc")
    exp = page.locator("#mem-res-edit-expiry")
    page.wait_for_function("document.querySelector('#mem-res-edit-desc').value === 'per-team memory'")
    check(exp.input_value() == "30", "inline form pre-filled from GET detail (desc + expiry 30)")
    save = page.get_by_test_id("memory-resource-save")
    hint = er.get_by_test_id("btn-hint")
    check(save.is_disabled() and hint.inner_text() == "Nothing changed yet.",
          f"SAVE disabled with reason before any change: {hint.inner_text()!r}")
    exp.fill("6")
    check(save.is_disabled() and "between 7 and 365" in hint.inner_text(),
          f"invalid expiry 6 keeps SAVE disabled with reason: {hint.inner_text()!r}")
    exp.fill("14")
    page.wait_for_function("!document.querySelector('[data-testid=memory-resource-save]').disabled")
    check(hint.count() == 0, "valid change enables SAVE and drops the hint")
    save.click()
    dlg = page.get_by_role("alertdialog")
    dlg.wait_for()
    text = dlg.inner_text()
    check("hr-assistant, faq-bot" in text and "(2)" in text, "confirm dialog lists the referencing agents")
    check("14 days" in text, "confirm dialog states the expiry impact on every agent")
    print("DIALOG:", text.replace("\n", " | "))
    page.screenshot(path=sys.argv[1] + "/confirm.png")
    dlg.locator(".confirm-actions button", has_text="SAVE").click()
    page.wait_for_function("!document.querySelector('[data-testid=memory-resource-edit-row]')")
    check(puts == [{"event_expiry_days": 14}], f"PUT body carries only the changed field: {puts}")
    rows = page.locator("tbody tr")
    check("UPDATING" in rows.nth(1).inner_text(), "row refreshed from the PUT response (status UPDATING)")
    check(len(list_calls) == 1, f"no list reload after save (list GETs = {len(list_calls)})")
    check(page.locator("text=Memory team_notes-XYZ789 updated.").count() >= 1, "success toast shown")
    page.screenshot(path=sys.argv[1] + "/after-save.png")
    # zh-CN copy renders too
    page.evaluate("localStorage.setItem('i18nextLng','zh-CN')")
    page.goto(f"{BASE}/memory?view=resources")
    page.get_by_test_id("memory-resource-edit").nth(0).wait_for()
    check(page.get_by_test_id("memory-resource-edit").nth(0).inner_text().strip() == "编辑",
          "zh-CN EDIT label")
    print("unhandled stub paths:", unhandled)
    b.close()
print("ALL PASS")

File: self-evolution/runs/SE-013/ui_probe.mjs
// SE-013 UI probe: RATE LIMITS panel against a stubbed /api (no backend, no AWS).
// Usage: node ui_probe.mjs http://127.0.0.1:5199
import { chromium } from "/home/ubuntu/.nvm/versions/node/v22.19.0/lib/node_modules/playwright/index.mjs";

const BASE = process.argv[2] ?? "http://127.0.0.1:5199";
const OUT = "/home/ubuntu/workspace/agentcore_launchpad/.claude/self-evolution/runs/SE-013";
const results = [];
const check = (name, ok, detail = "") => {
  results.push({ name, ok, detail });
  console.log(`${ok ? "PASS" : "FAIL"} ${name}${detail ? ` — ${detail}` : ""}`);
};

const gateway = {
  id: "gw-1", arn: "arn:aws:bedrock-agentcore:us-west-2:123:gateway/gw-1", name: "launchpad-gw",
  description: "", status: "READY", status_reasons: [], protocol_type: "MCP",
  authorizer_type: "AWS_IAM", url: "https://gw-1.example.test/mcp", role_arn: "arn:aws:iam::123:role/gw",
  managed: true, target_count: 0, targets: [], policy_engine: null,
  shared_gateways: [{ id: "gw-1", arn: "arn", name: "launchpad-gw" }], shared_engine: false,
  attachability: { attachable: true, reason: null, auth_type: "aws_iam" }, policy_test_available: false,
  registry_record: null, legacy_record_count: 0, updated_at: "2026-09-05T00:00:00+00:00",
  authorizer_configuration: null, protocol_configuration: null, actions: [], iam_preflight: null,
  external_tools_list_command: null,
};
const limit = (id, extra = {}) => ({
  id, gateway_id: "gw-1", description: "per target", dimension_keys: ["targetName", "$.context.jwt.sub"],
  entries: [
    { dimensions: { targetName: "office-facts", "$.context.jwt.sub": "*" },
      requests: [{ rate: 10, period: "second" }], tokens: [{ rate: 5000, period: "minute" }] },
    { dimensions: { targetName: "*", "$.context.jwt.sub": "*" }, requests: [{ rate: 60, period: "minute" }] },
  ],
  status: "ACTIVE", created_at: "2026-09-05T00:00:00+00:00", updated_at: "2026-09-05T01:02:03+00:00", ...extra,
});

const state = { rateLimits: { status: 200, body: { rate_limits: [limit("rl-1"), limit("rl-2", { status: "UPDATING" })] } }, posts: [] };
const browser = await chromium.launch();
const page = await browser.newPage({ viewport: { width: 1500, height: 1000 } });
await page.addInitScript(() => localStorage.setItem("i18nextLng", "en"));
const unhandled = new Set();
await page.route("**/api/**", async (route) => {
  const req = route.request();
  const url = new URL(req.url());
  const p = url.pathname;
  const json = (status, body) => route.fulfill({ status, contentType: "application/json", body: JSON.stringify(body) });
  if (p === "/api/auth/status") return json(200, { auth_required: false, authenticated: true, registration_enabled: false, registration_requires_approval: false, username: null, role: "admin", email: null, account_expires_at: null, permissions: [] });
  if (p === "/api/governance/gateways/gw-1") return json(200, gateway);
  if (p === "/api/governance/gateways/gw-1/policies") return json(200, { policies: [] });
  if (p === "/api/governance/gateways/gw-1/registry-preview") return json(503, { code: "registry.unavailable", message: "stub", detail: null });
  if (p === "/api/governance/gateways/gw-1/decisions") return json(200, { available: true, log_only_count: 0, decisions: [] });
  if (p === "/api/governance/gateways/gw-1/rate-limits") {
    if (req.method() === "POST") {
      state.posts.push(JSON.parse(req.postData()));
      return json(201, limit("rl-new", { status: "CREATING" }));
    }
    return json(state.rateLimits.status, state.rateLimits.body);
  }
  if (p.startsWith("/api/governance/gateways/gw-1/rate-limits/")) {
    state.posts.push({ method: req.method(), path: p, body: req.postData() ? JSON.parse(req.postData()) : null });
    return json(200, req.method() === "DELETE" ? { deleted: true, id: "rl-1", status: "DELETING" } : limit("rl-1", { status: "UPDATING" }));
  }
  unhandled.add(`${req.method()} ${p}`);
  return json(404, { code: "http.404", message: "stub", detail: null });
});

const go = () => page.goto(`${BASE}/governance?view=gateway&gateway=gw-1`, { waitUntil: "networkidle" });
const panel = () => page.locator(".panel", { has: page.locator("h2", { hasText: "RATE LIMITS" }) });

// ── 1. list ──────────────────────────────────────────────────────────────────
await go();
await panel().waitFor();
const rows = await panel().locator('tr[data-testid^="rate-limit-"]').count();
check("list renders two rate limits", rows === 2, `rows=${rows}`);
check("status chips", (await panel().innerText()).includes("ACTIVE") && (await panel().innerText()).includes("UPDATING"));
check("semantics note visible", (await panel().innerText()).includes("evaluated before Policy"));
await panel().locator('tr[data-testid="rate-limit-rl-1"] .gov-rl-toggle').click();
const entriesText = await panel().locator(".gov-rl-entries").innerText();
check("expanded entries show dimension values + metric rates", entriesText.includes("office-facts") && entriesText.includes("requests 10/second") && entriesText.includes("tokens 5000/minute"), entriesText.replace(/\s+/g, " ").slice(0, 160));
const editDisabledReason = await panel().locator('tr[data-testid="rate-limit-rl-2"] button:has-text("EDIT ENTRIES")').getAttribute("title");
check("EDIT on UPDATING row disabled with reason", (editDisabledReason ?? "").includes("UPDATING"), editDisabledReason ?? "");
await page.screenshot({ path: `${OUT}/ui-list.png`, fullPage: true });

// ── 2. empty ─────────────────────────────────────────────────────────────────
state.rateLimits = { status: 200, body: { rate_limits: [] } };
await go();
await panel().waitFor();
check("empty state", (await panel().innerText()).includes("NO RATE LIMITS ON THIS GATEWAY"));

// ── 3. error ─────────────────────────────────────────────────────────────────
state.rateLimits = { status: 403, body: { code: "aws.access_denied", message: "denied", detail: null } };
await go();
await panel().waitFor();
await page.waitForTimeout(300);
const errText = await panel().innerText();
check("error state shows the envelope, not an empty table", errText.includes("aws.access_denied") && !errText.includes("NO RATE LIMITS ON THIS GATEWAY"), errText.replace(/\s+/g, " ").slice(0, 200));
await page.screenshot({ path: `${OUT}/ui-error.png`, fullPage: true });

// ── 4. ADD form: client-side validation ──────────────────────────────────────
state.rateLimits = { status: 200, body: { rate_limits: [] } };
await go();
await panel().waitFor();
await panel().locator('button:has-text("ADD RATE LIMIT")').click();
const form = page.locator('[data-testid="rate-limit-form"]');
await form.waitFor();
const save = form.locator('button:has-text("SAVE")');
check("SAVE disabled on empty form with reason", (await save.isDisabled()) && ((await save.getAttribute("title")) ?? "").includes("dimension key"), (await save.getAttribute("title")) ?? "");
await form.locator('.gov-rl-key-picker button:has-text("targetName")').click();
await form.locator('.gov-rl-key-picker button:has-text("toolName")').click();
await form.locator('button:has-text("ADD ENTRY")').click();
const entry = form.locator('[data-testid="rate-limit-entry-0"]');
const dims = entry.locator(".gov-rl-dimension input");
await dims.nth(0).fill("*");
await dims.nth(1).fill("search");
await entry.locator('.gov-rl-metric input[type="checkbox"]').nth(0).check();
await entry.locator('input[aria-label="requests RATE"]').fill("5");
let blockers = await form.locator('[data-testid="rate-limit-blockers"]').innerText();
check("leading * is blocked client-side", (await save.isDisabled()) && blockers.includes("trailing"), blockers);
check("SAVE explains why (disabledReason)", ((await save.getAttribute("title")) ?? "").includes("trailing"));
await dims.nth(0).fill("office-facts");
await entry.locator('.gov-rl-metric input[type="checkbox"]').nth(1).check(); // tokens
await entry.locator('input[aria-label="tokens RATE"]').fill("1000");
await entry.locator('select[aria-label="tokens PERIOD"]').selectOption("second");
blockers = await form.locator('[data-testid="rate-limit-blockers"]').innerText();
check("tokens per second is blocked client-side", (await save.isDisabled()) && blockers.includes("tokens are per minute only"), blockers);
await entry.locator('select[aria-label="tokens PERIOD"]').selectOption("minute");
await page.waitForTimeout(100);
check("valid draft enables SAVE", !(await save.isDisabled()) && (await form.locator('[data-testid="rate-limit-blockers"]').count()) === 0);
// jwt claim key with invalid then valid claim
await form.locator('input[aria-label="JWT CLAIM"]').fill("a");
check("1-char JWT claim cannot be added", await form.locator('button:has-text("ADD KEY")').isDisabled());
await form.locator('input[aria-label="JWT CLAIM"]').fill("sub");
await form.locator('button:has-text("ADD KEY")').click();
check("jwt key added and entry gained a * for it", (await form.locator('[data-testid="rate-limit-keys"]').innerText()).includes("$.context.jwt.sub") && (await dims.count()) === 3 && (await dims.nth(2).inputValue()) === "*");
await form.locator('input.input:not(.mono)').last().fill("probe description");
await page.screenshot({ path: `${OUT}/ui-form.png`, fullPage: true });
await save.click();
await page.waitForTimeout(500);
const post = state.posts.find((p) => p.dimension_keys);
check("POST body is the wire shape", !!post && JSON.stringify(post) === JSON.stringify({
  dimension_keys: ["targetName", "toolName", "$.context.jwt.sub"],
  entries: [{ dimensions: { targetName: "office-facts", toolName: "search", "$.context.jwt.sub": "*" }, requests: [{ rate: 5, period: "second" }], tokens: [{ rate: 1000, period: "minute" }] }],
  description: "probe description",
}), JSON.stringify(post));
check("form closes after save", (await page.locator('[data-testid="rate-limit-form"]').count()) === 0);

// ── 5. edit keeps keys locked; delete confirms ───────────────────────────────
state.rateLimits = { status: 200, body: { rate_limits: [limit("rl-1")] } };
await go();
await panel().waitFor();
await panel().locator('tr[data-testid="rate-limit-rl-1"] button:has-text("EDIT ENTRIES")').click();
await form.waitFor();
check("edit locks keys (no picker, lock hint)", (await form.locator(".gov-rl-key-picker").count()) === 0 && (await form.innerText()).includes("immutable"));
check("edit hydrates two entries", (await form.locator('[data-testid^="rate-limit-entry-"]').count()) === 2);
await save.click();
await page.waitForTimeout(400);
const put = state.posts.find((p) => p.method === "PUT");
check("PUT sends entries + description only", !!put && Object.keys(put.body).sort().join(",") === "description,entries", JSON.stringify(put?.body).slice(0, 200));
await panel().locator('tr[data-testid="rate-limit-rl-1"] button:has-text("DELETE")').click();
await page.locator('[role="alertdialog"]').waitFor();
check("delete confirm names the rate limit", (await page.locator('[role="alertdialog"]').innerText()).includes("rl-1"));
await page.locator('[role="alertdialog"] button:has-text("DELETE")').click();
await page.waitForTimeout(400);
check("DELETE request sent", state.posts.some((p) => p.method === "DELETE" && p.path.endsWith("/rl-1")));

// ── 6. unmanaged gateway: ADD disabled with reason ───────────────────────────
gateway.managed = false;
await go();
await panel().waitFor();
const addTitle = await panel().locator('button:has-text("ADD RATE LIMIT")').getAttribute("title");
check("unmanaged → ADD disabled with 'manage this Gateway first'", (addTitle ?? "").includes("manage this Gateway first"), addTitle ?? "");

// ── 7. zh-CN renders ─────────────────────────────────────────────────────────
await page.evaluate(() => localStorage.setItem("i18nextLng", "zh-CN"));
await page.addInitScript(() => localStorage.setItem("i18nextLng", "zh-CN"));
await go();
await page.waitForTimeout(300);
check("zh-CN panel title", (await page.locator(".panel h2", { hasText: "限流" }).count()) === 1);

console.log("unhandled api:", [...unhandled].join(", ") || "none");
await browser.close();
const failed = results.filter((r) => !r.ok);
console.log(`\n${results.length - failed.length}/${results.length} checks passed`);
process.exit(failed.length ? 1 : 0);
